Make π < π work as expected

perrutquist opened the pull request:
Methods are added for < and <= comparing Irrationals. (The corresponding
methods for > and >= are then handled by existing fallback methods.)

Currently, expressions like `ℯ < π` result in promotion to Float64.
This works most of the time. However, `π < π` does not result in
promotion and throws an error. This commit introduces methods
for comparing an Irrational to itself.

This commit also introduces a test to make sure that incorrect results
are not being returned due to different user-defined irrationals
rounding to the same Float64.

perrutquist committed:
Make π < π work as expected

Methods are added for < and <= comparing Irrationals. (The corresponding
> and >= are then handled by existing fallback methods.)

Currently, expressions like `ℯ < π` result in promotion to Float64.
This works most of the time. However, `π < π` does not result in
promotion and throws an error. This commit introduces a method
for comparing an Irrational to itself.

This commit also introduces a test to make sure that incorrect results
are not being returned due to different user-defined irrationals
rounding to the same Float64.

martinholters commented on base/irrationals.jl:
```diff
@@ -56,6 +56,17 @@ float(::Type{<:AbstractIrrational}) = Float64
+function <(x::AbstractIrrational, y::AbstractIrrational)
+    # If two different Irrationals round to the same Float64, then explicit
+    # < methods are needed for those inputs, as the below assertation will fail.
+    @assert Float64(x) != Float64(y)
```
This should probably throw some specific error.

simonbyrne replied:
Probably a good idea. Something like
```
Float64(x) != Float64(y) || error("$x and $y have identical Float64 representations")
```

martinholters replied:
Or a `MethodError`. The solution is to add a specific method, after all.

perrutquist replied:
I've added a commit, changing the code to throw a `MethodError`.

simonbyrne commented:
I thought it did, but I just tried various versions back to 0.4 and apparently not!

perrutquist committed:
MethodError instead of AssertionError from <

In the (probably rare) case where a user defines two Irrationals that do
not differ in their Float64 representation, comparisons will now
throw a MethodError to inform the user that he needs to define the <
method himself.

waldyrious commented:
Cross-referencing the Discourse thread where this issue was raised: https://discourse.julialang.org/t/comparison-of-same-irrationals-by-throws-error/1342/
